### ISSUE TITLE
Future problem of a-asset-item and gltf-model combination

### ISSUE THREAD

**Issue by takahirox** [aframevr/aframe#2378]
**Description:**

I've run into this problem while I was trying to bump Three.js to r84.

In r84 `THREE.GLTFLoader` expects that glTF file is loaded as arraybuffer.
(In r83 it loads as text)

https://github.com/mrdoob/three.js/blob/r84/examples/js/loaders/GLTFLoader.js#L28

But `a-asset-item` loads data as DOMString(text) in advance and keeps it in `THREE.Cache`.

`THREE.FileLoader` returns cached data if cache hits.

I mean, `THREE.FileLoader` returns cached text even though
`THREE.GLTFLoader` sets `THREE.FileLoader.setResponseType('arraybuffer')`
if `a-asset-item` loads in advance.

Actually `gltf-model.test` fails because of this issue with Three.js r84.

I have some ideas in my mind to solve this problem.

1. We don't use `a-asset-item` for `gltf-model`. Easy but inconvenient.
2. Disable `THREE.Cache`. Easy but worse performance.
3. Enable arraybuffer `a-asset-item` (#2112). (Add responseType prop?)
4. `a-asset-item` checks the extension of file and loads as arraybuffer if it's .gltf or .glb

I prefer 3.
Which one do you guys like or any other better ideas?

/cc @donmccurdy

**Comment by ngokevin** [comment]
3. sounds good to me.

**Comment by donmccurdy** [comment]
[Context for the change in r84](https://github.com/mrdoob/three.js/pull/10500#issuecomment-270354703).

I also like (3). Should responseType be set automatically, or do we just document that it needs to be done for glTF? 

GLTFLoader has an internal cache, sort of, so disabling THREE.Cache for `gltf-model` only might be a 5th option.

**Comment by takahirox** [comment]
IMO, it'd be nice to have 3. and 4.

We provide a way to set responseType in `a-asset-item` like this

    <a-asset-item id="model" responseType="arraybuffer" src="model.gltf"></a-asset-item>

It'd be worth for A-Frame non-core components which need to load data as non-text.
(`a-mmd` I've making is actually one of them)
They document the data type.

And also `a-asset-item` automatically sets appropriate responseType for the data type which
A-Frame core supports. (arraybuffer for .gltf and .glb)

**Comment by takahirox** [comment]
I'm working on updating `a-asset-item` and have two questions.

1. Why does `a-asset-item` call `THREE.Cache.files[src] = responseText;` 

https://github.com/aframevr/aframe/blob/d055cfc56ef263d69c89c266921d6d9c4f5c3691/src/core/a-assets.js#L102

although it's called in `THREE.FileLoader`?

https://github.com/mrdoob/three.js/blob/accd0b3b2067ac05cd2557252ba4394fd04c7939/src/loaders/FileLoader.js#L137

And if it needs to explicitly call, why doesn't it use `THREE.Cache.add(src, responseText);`?

2. In `a-assets.test.js` this `var XHR_SRC = 'base/src/README.md';` is the path to non-exist file.

https://github.com/aframevr/aframe/blob/96df6c1f22fdf72c392be710d945bb9e8684b05c/tests/core/a-assets.test.js#L8

Is this intended?

**Comment by donmccurdy** [comment]
> 1. Why does a-asset-item call THREE.Cache.files[src] = responseText;

I have no idea — seems fine to change that, as long as we set `THREE.Cache.enabled = true` somewhere.

> 2. In a-assets.test.js this var XHR_SRC = 'base/src/README.md'; is the path to non-exist file.

I also have no idea. 😑

**Comment by ngokevin** [comment]
1. was from https://github.com/aframevr/aframe/commit/5728902d29cc73b12269f41d5417ae92ab63ae77

You can try to remove, I don't quite remember, but I might have been running into issues where it didn't exist in the cache in time the A-Frame code ran.

**Comment by takahirox** [comment]
I tried. The all test passed. I think I can remove.

Just in case, I checked if cache works with primitives/model example.

![image](https://cloud.githubusercontent.com/assets/7637832/23336404/244bcbb6-fb84-11e6-9214-4cf3a6d58baf.png)

1. before removing (ignore .dae because `THREE.ColladaLoader` doesn't support `THREE.Cache` in r83)
![image](https://cloud.githubusercontent.com/assets/7637832/23336396/d38d9128-fb83-11e6-925d-6532e628e896.png)

2. after removing. Still asstes are downloaded only once.
![image](https://cloud.githubusercontent.com/assets/7637832/23336395/b22170ea-fb83-11e6-85e4-57a95468f31d.png)

3. after I remove and comment out `THREE.Cache.enabled=true`. Assets are downloaded twice. 

https://github.com/aframevr/aframe/blob/6306459b2d193d7951e52ffb10bdfb3c10cdbea8/src/lib/three.js#L18

![image](https://cloud.githubusercontent.com/assets/7637832/23336400/e76fff46-fb83-11e6-8df0-b90d0a9ce3cf.png)

**Comment by takahirox** [comment]
And I guess

>In a-assets.test.js this var XHR_SRC = 'base/src/README.md'; is the path to non-exist file.

is a bug.

I'll make PR to fix them.

**Comment by takahirox** [comment]
BTW, what's this line in src/lib/three.js for?

    if (THREE.Cache) {

https://github.com/aframevr/aframe/blob/6306459b2d193d7951e52ffb10bdfb3c10cdbea8/src/lib/three.js#L17
